pkg/exchange/max/exchange.go

```diff
@@ -1100,6 +1089,76 @@ func (e *Exchange) QueryTrades(
 	return trades, nil
 }
 
+// prepareQueryTradesRequest helps to prepare the query parameters for trades request.
+// If we use timestamp as parameter, MAX will ignore from_id.
+func (e *Exchange) prepareQueryTradesRequest(
+	ctx context.Context, symbol string, options *types.TradeQueryOptions,
+	req *v3.GetWalletTradesRequest,
+) {
+	if options.LastTradeID != 0 {
+		// the caller try to query trades that are after the given trade ID
+		lastTrade, err := e.queryTradeByID(ctx, symbol, options.LastTradeID)
+		if err != nil {
+			log.WithError(err).Infof("failed to query last trade with id %d, fallback to time range query", options.LastTradeID)
+			// last trade not found, just use time range for query
+			if options.StartTime != nil {
+				req.Timestamp(*options.StartTime)
+				req.Order("asc")
+			} else if options.EndTime != nil {
+				req.Timestamp(*options.EndTime)
+				req.Order("desc")
+			}
+			return
+		}
+		if options.StartTime != nil && lastTrade.Time.After(*options.StartTime) {
+			// the last trade is after the given start time -> use from_id instead
+			log.Infof("the last trade %d time %s is after %s, use from_id for query", options.LastTradeID, lastTrade.Time, options.StartTime)
+			req.FromID(options.LastTradeID)
+			req.Order("asc")
+			return
+		}
+	}
+	// we can only use time range for query from here
+	if options.StartTime != nil {
+		req.Timestamp(*options.StartTime)
+		req.Order("asc")
+	} else if options.EndTime != nil {
+		req.Timestamp(*options.EndTime)
+		req.Order("desc")
+	}
+}
+
+func (e *Exchange) queryTradeByID(ctx context.Context, symbol string, tradeID uint64) (*types.Trade, error) {
+	if err := e.queryTradeLimiter.Wait(ctx); err != nil {
+		return nil, err
+	}
+
+	market := toLocalSymbol(symbol)
+	walletType := maxapi.WalletTypeSpot
+	if e.MarginSettings.IsMargin {
+		walletType = maxapi.WalletTypeMargin
+	}
+	req := e.v3client.NewGetWalletTradesRequest(walletType)
+	req.Market(market).Limit(1)
+	req.FromID(tradeID)
+	maxTrades, err := req.Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(maxTrades) == 0 {
+		return nil, fmt.Errorf("trade %d not found", tradeID)
```

Comment:
for new strategy setup, it's possible we found no trades, and that's normal and expected

Reply:
In that case, we will fallback to time range only query, which I think is quite reasonable.